File: reference-apps/nextjs-postgres-crud/lib/guard.ts
import { NextResponse } from "next/server";
import { getSession, type SessionPayload } from "@/lib/auth";
import { log } from "@/lib/log";

// access-control guard: every API route that touches a task calls this first.
// Ownership/role checks live here so app/api routes stay declarative — see
// claude-setup/tests/fixtures/profiles/internal-crud.json ("access-control": required).
export async function requireSession(): Promise<{ session: SessionPayload } | { error: NextResponse }> {
  const session = await getSession();
  if (!session) {
    return { error: NextResponse.json({ error: "unauthenticated" }, { status: 401 }) };
  }
  return { session };
}

export function canAccessTask(session: SessionPayload, task: { ownerId: string }): boolean {
  return session.role === "admin" || task.ownerId === session.sub;
}

export function forbidden(session: SessionPayload, taskId: string): NextResponse {
  // A 403 on an authenticated request is a real ownership-boundary event worth an audit trail of
  // its own, distinct from the per-record AuditLog table (lib/audit.ts) which only records changes
  // that were actually allowed to happen.
  log.warn("access.forbidden", { userId: session.sub, role: session.role, taskId });
  return NextResponse.json({ error: "forbidden" }, { status: 403 });
}

File: reference-apps/react-fastapi-postgres-crud/backend/app/guard.py
"""access-control guard: every API route that touches a task calls require_session first.

Ownership/role checks live here so router functions stay declarative — see
claude-setup/tests/fixtures/profiles/internal-crud.json ("access-control": required).
"""

from fastapi import Cookie, HTTPException, Request

from app.auth import SESSION_COOKIE_NAME, SessionPayload, decode_session
from app.log import log
from app.models import Task


def require_session(session: str | None = Cookie(default=None, alias=SESSION_COOKIE_NAME)) -> SessionPayload:
    payload = decode_session(session)
    if payload is None:
        raise HTTPException(status_code=401, detail="unauthenticated")
    return payload


def can_access_task(session: SessionPayload, task: Task) -> bool:
    return session["role"] == "admin" or task.owner_id == session["sub"]


def forbidden(session: SessionPayload, task_id: str, request: Request | None = None) -> HTTPException:
    # A 403 on an authenticated request is a real ownership-boundary event worth an audit
    # trail of its own, distinct from the per-record AuditLog table (app/audit.py) which
    # only records changes that were actually allowed to happen.
    log.warn("access.forbidden", {"userId": session["sub"], "role": session["role"], "taskId": task_id})
    return HTTPException(status_code=403, detail="forbidden")
